Add Kandinsky 5 Pro LoRA training

AkaneTendo25 opened the pull request:
@kohya-ss 

Sorry for submitting yet another PR before finishing my previous one ([for LongCat-Video](https://github.com/kohya-ss/musubi-tuner/pull/712)), I will take care of the latter as well soon. But I wanted to be the first to contribute K5 training pipeline to your wonderful project ❤️
I will not abandon this PR and plan to continue pushing it toward production readiness.

The Kandinsky-5 Pro model is promising, and thanks to the ComfyUI team ([especially Kijai](https://github.com/comfyanonymous/ComfyUI/pull/10988)) it should soon be available to a wide audience in the main ComfyUI release (and probably should run fast enough with distillation LoRA). I hope users will try it and will love to create some great LoRAs for it.

This is a draft version, and I've only tested T2V LoRA, but both the training and inference pipelines are working. With musubi-tuner optimizations (fp8 and/or block swapping), video training fits well within consumer setup, with tolerable (expected for a 19B model) performance. With 10 blocks swapped I get 5.5s/it with mixed (512p images + 256p/17f videos) dataset on Windows 11, 24 Gb VRAM (RTX 3090), DDR4 128 Gb RAM (64 Gb probably will be enough, but cannot confirm). Pretrain-5s model fine-tunes reliably and demonstrates strong style capture. 

Training and inference examples:

<details><summary>Details</summary>
<p>

VAE latents cache:
```
python kandinsky5_cache_latents.py ^
  --dataset_config G:/samples/kandinsky-tuner/_kandinsky_lora_dataset.toml ^
  --vae F:/k5_modules/hunyuan/vae/diffusion_pytorch_model.safetensors ^
  --device cuda ^
  --vae_dtype fp16
```

Text embeddings cache:
```
python kandinsky5_cache_text_encoder_outputs.py ^
  --dataset_config G:/samples/kandinsky-tuner/_kandinsky_lora_dataset.toml ^
  --text_encoder_qwen F:/k5_modules/text_encoder ^
  --text_encoder_clip F:/k5_modules/text_encoder2 ^
  --device cuda ^
  --quantized_qwen ^
  --batch_size 4
```

Training (T2V):
```
accelerate launch --num_cpu_threads_per_process 1 --mixed_precision bf16 kandinsky5_train_network.py ^
  --mixed_precision bf16 ^
  --dataset_config G:/samples/kandinsky-tuner/_kandinsky_lora_dataset.toml ^
  --task k5-pro-t2v-5s-sd ^
  --dit F:/kandinsky5pro_t2v_pretrain_5s.safetensors ^
  --text_encoder_qwen F:/k5_modules/text_encoder ^
  --text_encoder_clip F:/k5_modules/text_encoder2 ^
  --vae F:/k5_modules/hunyuan/vae/diffusion_pytorch_model.safetensors ^
  --fp8_base ^
  --blocks_to_swap 10 ^
  --flash_attn ^
  --gradient_checkpointing ^
  --max_data_loader_n_workers 1 ^
  --persistent_data_loader_workers ^
  --learning_rate 1e-4 ^
  --optimizer_type AdamW8Bit ^
  --optimizer_args "weight_decay=0.001" "betas=(0.9,0.95)" ^
  --max_grad_norm 1.0 ^
  --lr_scheduler constant_with_warmup ^
  --lr_warmup_steps 100 ^
  --network_module networks.lora_kandinsky ^
  --network_dim 16 ^
  --network_alpha 16 ^
  --timestep_sampling shift ^
  --discrete_flow_shift 5.0 ^
  --output_dir G:/samples/kandinsky-tuner/output ^
  --output_name kandinsky_lora_moebius ^
  --log_config ^
  --log_with tensorboard ^
  --logging_dir G:/samples/kandinsky-tuner/logs ^
  --save_state ^
  --save_every_n_epochs 5 ^
  --max_train_epochs 50 ^
  --sample_prompts G:/samples/kandinsky-tuner/_kandinsky_lora_sampling.txt ^
  --scheduler_scale 10.0 ^
  --offload_dit_during_sampling ^
  --sample_every_n_epochs 5
```

Generate sample video (T2V):
```
python kandinsky5_generate_video.py ^
  --task k5-pro-t2v-5s-sd ^
  --dit F:/kandinsky5pro_t2v_pretrain_5s.safetensors ^
  --vae F:/k5_modules/hunyuan/vae/diffusion_pytorch_model.safetensors ^
  --text_encoder_qwen F:/k5_modules/text_encoder ^
  --text_encoder_clip F:/k5_modules/text_encoder2 ^
  --blocks_to_swap 10 ^
  --offload_dit_during_sampling ^
  --fp8_base ^
  --dtype bfloat16 ^
  --prompt "Moebius style, woman with blonde hair walking on the beach, camera slowly zooms out." ^
  --negative_prompt "low quality, artifacts" ^
  --frames 17 ^
  --steps 10 ^
  --guidance 5 ^
  --scheduler_scale 10 ^
  --seed 42 ^
  --width 384 ^
  --height 384 ^
  --lora_weight G:/samples/kandinsky-tuner/output/kandinsky_lora_moebius-000010.safetensors ^
  --lora_multiplier 1.0 ^
  --output output\k5_sample_lora.mp4
```

</p>
</details> 

Roadmap (remaining tasks):

- [x] Implement I2V training
- [x] Fix incorrect AI-generated documentation
- [x] Add [NABLA](https://arxiv.org/abs/2507.13546v1) [sparse attention](https://github.com/gen-ai-team/Wan2.1-NABLA) for fine-tuning CrossDiT (see also [K5 tech report](https://arxiv.org/abs/2511.14993), section 5.4)
- [x] (optional) Verify pipeline compatibility with K5-Lite video models
- [x] <del>(optional) Add support for T2I and I2I (editing) models</del> currently not planned
- [x] Remove unused code
- [x] Resolve issues found during review

Training with NABLA has been implemented, but I haven't been able to practically validate it. NABLA's benefits are expected mainly at higher video resolutions and longer durations, which I can't currently run on my hardware.

kabachuha commented:
I wonder what can even be trained on `256p/17f` 😬

AkaneTendo25 commented:
> I wonder what can even be trained on `256p/17f` 😬

@kabachuha Here this was just a smoke test 🙂. In general, however, this resolution is fine for training motion LoRAs.
For example, I trained Wan using a combination of 256x144x81 and 512x288x17 clips, along with 976x544 images, on an RTX 3090 (without block swapping), and the results were decent.

AkaneTendo25 commented:
@kabachuha Tbh I haven't yet tested training a LoRA specifically for this method, but based on the architecture there's a good chance it can work. As for inference, I haven't thought about it yet, but if there's no dedicated node for this (like WanFirstLastFrameToVideo), I'll try to make my own.

kabachuha commented:
@AkaneTendo25 I tried this one (adding also the code that replaced the last latent of the videos with the end frame), but it didn't work 😭

```python
class Kandinsky5FirstLastFrameToVideo(io.ComfyNode):
    @classmethod
    def define_schema(cls):
        return io.Schema(
            node_id="Kandinsky5FirstLastFrameToVideo",
            category="conditioning/video_models",
            inputs=[
                io.Conditioning.Input("positive"),
                io.Conditioning.Input("negative"),
                io.Vae.Input("vae"),
                io.Int.Input("width", default=768, min=16, max=nodes.MAX_RESOLUTION, step=16),
                io.Int.Input("height", default=512, min=16, max=nodes.MAX_RESOLUTION, step=16),
                io.Int.Input("length", default=121, min=1, max=nodes.MAX_RESOLUTION, step=4),
                io.Int.Input("batch_size", default=1, min=1, max=4096),
                io.Image.Input("start_image", optional=True),
                io.Image.Input("end_image", optional=True),
            ],
            outputs=[
                io.Conditioning.Output(display_name="positive"),
                io.Conditioning.Output(display_name="negative"),
                io.Latent.Output(display_name="latent", tooltip="Empty video latent"),
                io.Latent.Output(display_name="cond_latent_start", tooltip="Clean encoded start images, used to replace the noisy start of the model output latents"),
                io.Latent.Output(display_name="cond_latent_end", tooltip="Clean encoded end images, used to replace the noisy start of the model output latents"),
            ],
        )

    @classmethod
    def execute(cls, positive, negative, vae, width, height, length, batch_size, start_image=None, end_image=None) -> io.NodeOutput:
        latent = torch.zeros([batch_size, 16, ((length - 1) // 4) + 1, height // 8, width // 8], device=comfy.model_management.intermediate_device())
        cond_latent_start = {}
        cond_latent_end = {}
        conditioning_values = {}

        conditioning_values["concat_mask"] = torch.ones((1, 1, latent.shape[2], latent.shape[-2], latent.shape[-1]), device=latent.device, dtype=latent.dtype)

        if start_image is not None:
            start_image = comfy.utils.common_upscale(start_image[:length].movedim(-1, 1), width, height, "bilinear", "center").movedim(1, -1)
            encoded = vae.encode(start_image[:, :, :, :3])
            cond_latent_start["samples"] = encoded
            conditioning_values["concat_mask"][:, :, :((start_image.shape[0] - 1) // 4) + 1] = 0.0
            conditioning_values["time_dim_replace"] = encoded

        if end_image is not None:
            end_image = comfy.utils.common_upscale(end_image[:length].movedim(-1, 1), width, height, "bilinear", "center").movedim(1, -1)
            encoded = vae.encode(end_image[:, :, :, :3])
            cond_latent_end["samples"] = encoded
            conditioning_values["concat_mask"][:, :, -(((end_image.shape[0] - 1) // 4) + 1):] = 0.0
            conditioning_values["time_dim_replace_end"] = encoded

        positive = node_helpers.conditioning_set_values(positive, conditioning_values)
        negative = node_helpers.conditioning_set_values(negative, conditioning_values)

        out_latent = {}
        out_latent["samples"] = latent
        return io.NodeOutput(positive, negative, out_latent, cond_latent_start, cond_latent_end)
```

AkaneTendo25 commented:
@kohya-ss Sorry for this, I forgot to check generate video script after I changed attention mask processing during training (and this logic is shared by generate script).
Should be fixed now in https://github.com/kohya-ss/musubi-tuner/pull/774/commits/2897e17133a04471b9bdfa49e2f41d1d2f3a5dee

AkaneTendo25 commented:
Fixed it.

I also noticed a mistake in pyproject.toml: I added some scripts under src/musubi_tuner/kandinsky/ to the Ruff exception list, but the actual path is src/musubi_tuner/kandinsky**5**/.

Since all Kandinsky scripts now appear to be fully Ruff-compliant, would you prefer that I remove the Kandinsky-related exclusions from pyproject.toml altogether?

kohya-ss commented:
Thank you! If the scripts are already ruff-compliant, you can remove the exception from your pytproject.toml.

(I did that for HunyuanVideo 1.5.)
